library/kani_macros/src/sysroot/contracts.rs

```diff
@@ -966,3 +967,58 @@ macro_rules! passthrough {
 
 passthrough!(stub_verified, false);
 passthrough!(proof_for_contract, true);
+
+#[cfg(test)]
+mod test {
+    macro_rules! detect_impl_fn {
+        ($expect_pass:expr, $($tt:tt)*) => {{
+            let syntax = stringify!($($tt)*);
+            let ast = syn::parse_str(syntax).unwrap();
+            assert!($expect_pass == super::is_probably_impl_fn(&ast),
+                "Incorrect detection.\nExpected is_impl_fun: {}\nInput Expr; {}\nParsed: {:?}",
+                $expect_pass,
+                syntax,
+                ast
+            );
+        }}
+    }
+
+    #[test]
+    fn detect_impl_fn_by_receiver() {
+        detect_impl_fn!(true, fn self_by_ref(&self, u: usize) -> bool {});
+
+        detect_impl_fn!(true, fn self_by_self(self, u: usize) -> bool {});
+    }
+
+    #[test]
+    fn detect_impl_fn_by_self_ty() {
+        detect_impl_fn!(true, fn self_by_construct(u: usize) -> Self {});
+        detect_impl_fn!(true, fn self_by_wrapped_construct(u: usize) -> Arc<Self> {});
+
+        detect_impl_fn!(true, fn self_by_other_arg(u: usize, slf: Self) {});
+
+        detect_impl_fn!(true, fn self_by_other_wrapped_arg(u: usize, slf: Vec<Self>) {})
+    }
+
+    #[test]
+    fn detect_impl_fn_by_qself() {
+        detect_impl_fn!(
+            true,
+            fn self_by_mention(u: usize) {
+                Self::other(u)
```

Comment:
Is this possible given that the function doesn't have a return type?

Reply:
Of course. The whole issue is that you can have a function in an `impl` that possibly never mentions `self` or `Self`. But if you want to call that function (which we need to do in our generated code) you need to call it `Self::fn_name`.

for instance it's perfectly legal to do

```rs
struct S;
impl S {
  fn helper(u: usize) -> bool {
    u < 8
  }
}
```

But if you want to call it you have to do `S::foo(8)`.

That's why the whole self detector is just a heuristic and can actually lead to spurious compile errors. It's not very likely to have such a helper function but in can happen. What's worse actually is not using self when you should. For instance

```rs
impl<T> Vec<T> {
  fn new() -> Vec<T> {
    Vec { cap: 0, buf: NonNull::dangling() }
  }
}
```

This is perfectly valid Rust code , but wouldn't be detected by this as being in an `impl` because it uses `Vec<T>` as return type not `Self`. In the proc macro it's actually not possible to detect this case. And don't get fooled by the existence of the `ImplItem` enum and `ImplItemFn` struct, those you'll only get if you stick the macro on top of a whole `impl` block, then it gives you an `ItemImpl`, which contains them but if your macro is places on a function you can only parse an `ItemFn` (or possibly an `Item` but that will always be the `ItemFn` variant).

The basic problem with rust macros is that they only see the syntax that's given to them and no other context which is what you'd need to determine whether you're in an `impl`.

Reply:
OK, thanks for the explanation. I think it might be worth incorporating some of that explanation in the doc comments to make this clearer.

Reply:
I added the documentation I think you were referring to (and more). Let me know if anything else is missing.

Reply:
Looks great, thanks!